### PR TITLE
Finalized Torque Controller & Test Cases

### PR DESCRIPTION
**Pull request by SamAhnert**
@jcallaham With this addition we have the ability to input a torque command to control the Cylinder environment. This enables a more realistic action space for the RL agent to learn on.

### CODE REVIEW

**Review by jcallaham — CHANGES_REQUESTED** [review]
Nice work!  I think we will want to make a few changes before merging (details in the comments):

- Modify `IPCS` directly rather than having a new class
- Make sure the timestepper is abstracted away from details of the flow
- Consistent naming conventions
- Add tests into `test_cyl.py`

File: hydrogym/ts.py
Comment (SamAhnert) [reply hydrogym/ts.py:313]: 3.) yes that is correct, my instinct was to use fsolve for an implicit implementation of the solver (I believe that's what you mean when you say "non-autonomous ODE"), but after looking at the equation I think you're right that I can avoid using fsolve.

1&2.) The equation is:

`omega_t[i+1] = omega_t[i] + ((control_t[i+1] - k_damping*omega_t[i+1])/I_cm) * dt)`

which should be possible to solve without fsolve after a bit of rearranging.

Comment (SamAhnert) [reply hydrogym/ts.py:313]: after some simplification I get:

`self.omega = (self.omega + self.control * dt / self.I_cm) / (1 + self.k_damping * dt / self.Icm)`

which is maybe a bit less straight forward, but definitely faster. I can add some documentation to make sure it's clear where that comes from

File: hydrogym/ts.py
Comment (SamAhnert) [reply hydrogym/ts.py:294]: > Awesome that you found an actual value for this! Has this been nondimensionalized w.r.t. mass and diameter of the cylinder?

So following up on this point, is there a reason we are normalizing by diameter and not radius? 

Just asking because the equation is `I_cm = M R**2 / 2`, so would it make more sense to normalize by R? and whether or not it's D or R, should we normalize by (R/D)**2 ?

File: hydrogym/ts.py
Comment (SamAhnert) [reply hydrogym/ts.py:316]: The difference is in the implementation of `update_controls/update_velocity`.

`update_controls` assumes the control is the same as the state which is incompatible with the way `update_velocity` updates the state based on the control. If we were to merge IPCS_torque into IPCS, I believe this would only be possible via an indicator of whether torque or velocity is the control variable. Something like:
`
class IPCS(TransientSolver):
    def __init__(self, ..., control_variable='velocity', ...):
            super().__init__(flow, dt)
            ...
            self.control_variable = control_variable # either 'velocity' or 'torque'
`
Where later in the code we merge `update_controls/update_velocity` like so:

`def update_controls(self, control):
        if self.control_variable == 'torque':
            ...
            self.state = (self.state + self.control * dt / self.I_cm) / (1 + self.k_damping * dt / self.Icm)
        else:
            ...
            for (u, v) in zip(self.state, state):
                u = u + (self.dt/self.flow.TAU)*(v - u)

`

Then we would always just have `self.state` = `self.control` for the case where velocity is the control variable, which is a bit redundant but easy to do (or we could just ignore that self.state exists when using velocity as the control).

I do see the advantage to merging the two into one class though, so I will try and get something together over the next couple days while changing as little of the original code as possible.

Comment (SamAhnert) [reply hydrogym/ts.py:316]: @jcallaham let me know what you think of the above proposal to have an indicator on "what the control variable is". Also to make this more general, I should use other keywords besides 'torque' and 'velocity', but the general framework should be the same. 

Essentially, I would have a "directly controlling state_IPCS" where the control allows you to directly input the state that you want and an "input based_IPCS" similar to Torque_IPCS where the control is something that affects the current state of the environment indirectly. Then the "directly controlling state_IPCS" would be activated with one keyword and the "input based _IPCS" would be activated with a different keyword. I think this should work, but we'll see!

File: hydrogym/ts.py
Comment (jcallaham) [reply hydrogym/ts.py]: Either is fine in principle, but the rest of the flow configuration has been normalized with the length scale D (so that the cylinder has radius 0.5).  And I guess the mass scale should be set by the fluid density.  So I don't know what units the mass of this was in (if the radius was in inches, maybe lbs?), but start with the fluid density (I know this was a wind tunnel, but let's use water as the fluid here) in those units.  I'd convert everything to metric too, just to avoid wonky units.

Then rho=(mass)/(length)^3, and I=(mass)*(length)^2, so if `rho` is the density of water (in the correct units) and `D` is the diameter of the cylinder, I think you would just scale `I_cm` by `D**5/rho` to make it dimensionless... but check for yourself that that makes sense.

File: hydrogym/core.py
Comment (jcallaham) [inline hydrogym/core.py:205]: I think `state` might be a confusing name here, since `q = [u, p]^T` is what people will usually think of as the "state" of the flow.  Maybe `ctrl_state` or `actuator_state`?

File: hydrogym/ts.py
Comment (jcallaham) [inline hydrogym/ts.py:185]: `self.I_cm` and `self.k_damping` here have been renamed in lines 72-73, right?

File: hydrogym/ts.py
Comment (jcallaham) [inline hydrogym/ts.py:191]: This is nitpicky, but can we avoid the list `next_state` by just assigning `state = ...` within the loop?

File: hydrogym/ts.py
Comment (jcallaham) [inline hydrogym/ts.py:208]: Again, I think we should avoid having the timestepper depend on any details of the flow configuration.  For instance, if someone wanted to add their own flow configuration, they should just be able to implement a new subclass of `FlowConfig` without modifying the source of `ts.py` or anything else in the `hydrogym` package.

But it looks to me like this should be fine, assuming that these configurations provide methods like `get_damping` and `get_inertia`.  This looks like it should work with any number of inputs, which is the main thing.  I'd just maybe suggest a minor renaming to make it more general, since the equation for F=m*a also basically works for tau=I*\dot{omega} (I think this would also fix the comments above)

```
for (state, ctrl, m, k) in zip(
                    self.state, control, self.inertia, self.controller_damping_coef
                ):
    state = (state + ctrl * self.dt / m) / (1 + k* self.dt / m)
```

File: hydrogym/ts.py
Comment (jcallaham) [inline hydrogym/ts.py:231]: Do you have the pre-commit hooks Ludger added installed for formatting and linting?  If I remember correctly flake8 doesn't like the f-strings if there's no brackets in them (this was my fault originally but should be fixed now)

File: hydrogym/ts.py
Comment (jcallaham) [reply hydrogym/ts.py:208]: And maybe `get_damping` and `get_inertia` could raise a `NotImplementedError` similar to what you have here by default in `FlowConfig` until `Cavity`, `Pinball`, etc implement these.

File: hydrogym/ts.py
Comment (jcallaham) [inline hydrogym/ts.py:57]: I liked your original idea of "velocity" or "force" being the options - I think "direct" and "indirect" might be less obvious.

File: test/test_cyl.py
Comment (jcallaham) [inline test/test_cyl.py:144]: Naming convention here and below - `test_no_damp`, etc.

File: test/test_cyl.py
Comment (jcallaham) [inline test/test_cyl.py:171]: "Is reasonable" is pretty vague... is what reasonable?

**Review by jcallaham — CHANGES_REQUESTED** [review]
Looking good.  I like the backwards compatibility and all the tests a lot.  Just a few more things:

1. Make sure you have the pre-commit hooks installed (and let me know if you don't know how to do that).  There are a bunch of minor formatting/linting "changes" that should probably disappear with that.
2. The timestepping is much more abstracted from the flow config, but not fully (see comments on `update_controls`.  IPCS shouldn't care at all about what the flow is, or what its controls "mean".
3. I think as part of this upgrade we should add support at the `Env` level.  This should probably just be making sure that the args for control method get passed correctly, but you could add it to the unit tests (which would also have the side benefit of making the tests simpler and shorter).
4. I like how thorough you are with testing, but be mindful that we will probably be re-running these fairly often, so try to see if any of them can be consolidated or accelerated.  For instance, `test_shearForce0` could be incorporated into `test_steady`.  It's also fine to have more expansive local tests that you use for development but that we don't need to have as part of the full test suite to be re-run constantly as part of maintenance.  The key thing with the tests is to have quick checks of "I just changed A - did I accidentally break B?"

File: hydrogym/ts.py
Comment (jcallaham) [reply hydrogym/ts.py]: Yeah that's exactly the structure I had in mind.  Eventually we might even get rid of the "direct" version and just make sure that the default behavior is somehow equivalent (and that the "indirect" version is differentiable", but I like the backwards compatibility.  I actually liked "velocity" and "force" as keywords better than "direct" and "indirect" but otherwise I think this is a nice way of incorporating the functionality into the existing classes.

File: hydrogym/ts.py
Comment (SamAhnert) [reply hydrogym/ts.py:231]: oh crap, I thought we were doing black via #39 where Ludger says "Shall we go with black then? It does seem to be a pretty standard choice for most of the modern Python projects."

File: hydrogym/ts.py
Comment (SamAhnert) [reply hydrogym/ts.py]: Sure, I can do that. I do have some questions though.

The way I calculated I_cm was using some arbitrary numbers (density of plexiglass = 1.18 g/cm^3, length of cylinder = 1m). So by choosing these values, we are no longer simulating an arbitrary actuating cylinder, we are simulating a plexiglass cylinder with a length of 1 meter and a radius of 0.5. Since the radius of 0.5 m is built into the simulation, I don't see any problems there, but would it maybe be best to normalize the cylinder first by length and density of the plexiglass and then by the diameter? i.e. something like scaling `I_cm` by '1/(rho_plexiglass * L * D**4)' ? or are we okay with making the design choice that our cylinder will be of length 1m and "made" of plexiglass?

Also, I'm scaling by '1/(rho_plexiglass * L * D**4)' instead of the 'L*D**4/rho_plexiglass` format that you suggested since the fully expanded equation for I_cm would be `I_cm = (rho_plexiglass * L * pi * R**2) * R**2 / 2`

File: hydrogym/ts.py
Comment (SamAhnert) [reply hydrogym/ts.py:57]: My only concern is will this be abstract able to other Flow environments?

Comment (SamAhnert) [reply hydrogym/ts.py:57]: Actually I guess we could just throw all of "update_controls" into the flow config. Then, put this code block:

`
def update_controls(self, control):
            """Add a damping factor to the controller response

            If actual control is u and input is v, effectively
                du/dt = (1/tau)*(v - u)
            """
            for (u, v) in zip(self.control, control):
                u = u + (self.dt / self.flow.TAU) * (v - u)
            # for direct control of the state, the input is the desired state
            self.state = self.control
`

into the core.py so that all environments have access to the old way of updating state.

Then this could then be overridden in each individual environment so that we could use a keyword more environment specific. Then we would have update_environment in the cylinder env look like:

`
def update_controls(self, control):
        if self.control_method == "indirect":
            # Update state
            self.flow.update_state(control, self.dt)

        else:
            """Add a damping factor to the controller response

            If actual control is u and input is v, effectively
                du/dt = (1/tau)*(v - u)
            """
            for (u, v) in zip(self.control, control):
                u = u + (self.dt / self.flow.TAU) * (v - u)
            # for direct control of the state, the input is the desired state
            self.state = self.control
`

except with the appropriate variables called where necessary.


I will finish all other changes before coming back to this one since this seems like the most aggressive change.

Comment (SamAhnert) [reply hydrogym/ts.py:57]: PS idk why my code is formatting like that, but I hope the just of what I'm saying is clear without the examples

Comment (ludgerpaehler) [reply hydrogym/ts.py:57]: The other option for easier Python formatting is also to just use [YAPF](https://github.com/google/yapf) which will do its best to format your code in "good-faith" :)

Comment (SamAhnert) [reply hydrogym/ts.py]: So the more I thought about this, the more I believe that removing "update_controls/update_state" from TS and putting it in the flow config is necessary since (it is my understanding that) each flow environment will need to be controlled via it's own set of differential equations. Thus, "update_controls/update_state" is inherently unique to each flow environment.

I do understand that this is a pretty drastic change to the structure of the project, so if it interferes with development at the moment, we could push this to a later date and I can open a git issue with low priority. If this seems worth it to add now though, I was able to finish fully decoupling TS and Flow Config in my most recent push to this branch. All tests work for the cylinder and pinball. I'm just running into some issues with the "integrate" test in the cavity test suite which seems like a separate issue (via #38 ). Let me know what you think @jcallaham since I know you are still developing the backwards step and I don't want to throw a wrench in your work if it's too much for right now.

File: hydrogym/ts.py
Comment (jcallaham) [reply hydrogym/ts.py:164]: Oh yeah, it is black but also with flake8 through the [pre-commit](https://pre-commit.com/) config he set up. Do you have that installed? It should take care of all this automatically anytime you commit.  This was new to me as well, so let me know if you have any trouble getting it set up.

File: hydrogym/ts.py
Comment (jcallaham) [reply hydrogym/ts.py]: Oh alright, I was confused by the "2 inch radius" in the comment.  Yeah, you're definitely right the third dimension should be length and not radius though.  I'll have to work this out on paper I think, but let's leave it for now.  We can always update the numeric value if need be.  At the end of the day as long as it seems like it gives pretty good results I think it's fine.

So just to be clear for future reference, the properties you're assuming are for a 3D cylinder with density 1.18 g/cm^3 spanwise length 1m, and radius 0.5m?  And then `I_cm = (rho_plexiglass * L * pi * R**2) * R**2 / 2`?

File: hydrogym/ts.py
Comment (SamAhnert) [reply hydrogym/ts.py]: Awesome! This should be resolved then in commit "a4a6bf4" for the cylinder, pinball and cavity!

File: hydrogym/ts.py
Comment (SamAhnert) [reply hydrogym/ts.py]: Sorry about that, those comments are super confusing, and I'll clean them up in my next commit. I initially used other experimental papers as a point of reference, and they mostly used cylinders with a radius of ~2 inches so that's where that number comes from.

That equation is correct as well as the length and density. At the moment, I am using radius=0.5 according to the flow config, but I guess I can't say for sure that the radius is in units of meters. It's just given in the code as `self.rad = fd.Constant(0.5)` with no mention of units (or maybe R is intentionally left unitless?).

Comment (jcallaham) [reply hydrogym/ts.py]: Yeah it's meant to be dimensionless with respect to the diameter.  Okay I think this is good then - we can always tweak the numerical value if need be later on.

**Comment by jcallaham** [comment]
Oh, I think I see what the formatting issue is @SamAhnert.  I made a lot of these changes on `jc/testcase-validation` but I haven't merged them to main yet.  Let's do this - I'll merge #50 first, then let me know when you're all set with any other changes you want to make here and I can resolve whatever conflicts come up between this and `main`.  Then we can go ahead and merge this.

Definitely get `pre-commit` set up when you get a chance though.

**Comment by SamAhnert** [comment]
Sounds good, made my final changes and installed the precommit hooks. Should be ready to merge whenever!

**Review by jcallaham — APPROVED** [review]

**Pull request merged**
